cmd/admin/build_instance.go

```diff
@@ -0,0 +1,203 @@
+	}
+	return nil
+}
+
+func (b *buildInstanceCommand) writeJobResponseSecretToFile(
+	jobResponse *pb.RegisterInstanceResponse, dirPath string,
+) error {
+	// write all secrets into a file
+	secretsFileContent := ""
+	for key, val := range jobResponse.Context.Secrets {
+		if strings.Contains(val, unsubstitutedValue) {
+			b.keysWithUnsubstitutedValue = append(b.keysWithUnsubstitutedValue, key)
+		}
+		secretsFileContent += fmt.Sprintf("%s='%s'\n", key, val)
+	}
+
+	filePath := filepath.Join(dirPath, models.InstanceDataTypeSecretFileName)
+	writeToFileFn := utils.WriteStringToFileIndexed()
+	if err := writeToFileFn(filePath, secretsFileContent, b.logger.Writer()); err != nil {
+		return fmt.Errorf("failed to write asset file at %s: %w", filePath, err)
+	}
+	return nil
+}
+
+func (b *buildInstanceCommand) writeJobResponseEnvToFile(
+	jobResponse *pb.RegisterInstanceResponse, dirPath string,
+) error {
+	envFileBlob := ""
+	for key, val := range jobResponse.Context.Envs {
+		if strings.Contains(val, unsubstitutedValue) {
+			b.keysWithUnsubstitutedValue = append(b.keysWithUnsubstitutedValue, key)
+		}
+		envFileBlob += fmt.Sprintf("%s='%s'\n", key, val)
+	}
+
+	filePath := filepath.Join(dirPath, models.InstanceDataTypeEnvFileName)
+	writeToFileFn := utils.WriteStringToFileIndexed()
+	if err := writeToFileFn(filePath, envFileBlob, b.logger.Writer()); err != nil {
+		return fmt.Errorf("failed to write asset file at %s: %w", filePath, err)
+	}
+	return nil
+}
+
+func (b *buildInstanceCommand) writeJobResponseMapToFiles(
+	jobResponse *pb.RegisterInstanceResponse, dirPath string,
+) error {
+	if err := os.MkdirAll(dirPath, 0o777); err != nil {
+		return fmt.Errorf("failed to create directory at %s: %w", dirPath, err)
+	}
+
+	writeToFileFn := utils.WriteStringToFileIndexed()
+	for fileName, fileContent := range jobResponse.Context.Files {
+		filePath := filepath.Join(dirPath, fileName)
+		if err := writeToFileFn(filePath, fileContent, b.logger.Writer()); err != nil {
+			return fmt.Errorf("failed to write asset file at %s: %w", filePath, err)
+		}
+	}
+	return nil
+}
+
+func (b *buildInstanceCommand) sendInstanceRequest(jobName string, jobScheduledTimeProto *timestamppb.Timestamp) (*pb.RegisterInstanceResponse, error) {
+	conn, err := connectivity.NewConnectivity(b.clientConfig.Host, adminBuildInstanceTimeout)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	// fetch Instance by calling the optimus API
+	jobRun := pb.NewJobRunServiceClient(conn.GetConnection())
+	request := &pb.RegisterInstanceRequest{
+		ProjectName:  b.clientConfig.Project.Name,
+		JobName:      jobName,
+		ScheduledAt:  jobScheduledTimeProto,
+		InstanceType: pb.InstanceSpec_Type(pb.InstanceSpec_Type_value[utils.ToEnumProto(b.runType, "type")]),
+		InstanceName: b.runName,
+	}
+	return jobRun.RegisterInstance(conn.GetContext(), request)
+}
+
+func (b *buildInstanceCommand) getJobScheduledTimeProto() (*timestamppb.Timestamp, error) {
```

Comment:
We are not changing the struct, we should not be using pointer unless struct is big.